### ISSUE TITLE
how does gradient accumulate work in fairseq？

### ISSUE THREAD

**Issue by leon2milan** [pytorch/fairseq#3506]
I read the code.  I found accumulate gradients connect with the parameter of `update_freq`.  
In optimizer config, there is a update_fr eq config.  but I haven't found anything except in fp16optimizer.  
it shouldn't only support for fp16. 
So I search more, and found in model there's a parameter called update_num.  it maybe connect with criterion.  
for now, I still can't find any real implement about it.  
can anyone tell me. where I can get the information?

**Comment by lematt1991** [comment]
`update_freq` controls how many batches are accumulated before updating the parameters and clearing the gradients.

**Comment by leon2milan** [comment]
@lematt1991  yeah， I know.  But how?  where is the control code?

**Comment by lematt1991** [comment]
We create a `GroupedIterator` [here](https://github.com/pytorch/fairseq/blob/master/fairseq_cli/train.py#L243), which produces `update_freq` batches in each iteration.  [`train_step`](https://github.com/pytorch/fairseq/blob/master/fairseq/trainer.py#L636) then does a forward/backward pass for each batch in the group (accumulating gradients), and then takes a step after each batch in the group has been processed.

**Comment by leon2milan** [comment]
And then call `task.train_step`.  Use `criterion` calculate loss. But how to get loss in `accumulating gradients` ways. 
You know, general code like this :
```
model.zero_grad()                                   # Reset gradients tensors
for i, (inputs, labels) in enumerate(training_set):
    predictions = model(inputs)                     # Forward pass
    loss = loss_function(predictions, labels)       # Compute loss function
    loss = loss / accumulation_steps                # Normalize our loss (if averaged)
    loss.backward()                                 # Backward pass
    if (i+1) % accumulation_steps == 0:             # Wait for several backward steps
        optimizer.step()                            # Now we can do an optimizer step
        model.zero_grad()
```

**Comment by lematt1991** [comment]
Sorry, I'm not sure I understand the question.  Are you asking how to return the loss from `train_step`?

**Comment by leon2milan** [comment]
I'm asking how to return the loss with accumulating gradients.   How it works?
Accumulating gradients means after accumulation_steps or accumulation_steps step, we call `optimizer.step()` and `model.zero_grad()` to  update parameter.  Right?
I cannot find the control code like this in fairseq.   In many criterions, they just return the loss without accumulated normalization.
